### PR TITLE
Metrics tests

### PR DESCRIPTION
**Pull request by kvahed**
Adds more tests for metrics feature.
Documentation PR: https://github.com/arangodb/docs/pull/252

### CODE REVIEW

**Comment by kvahed** [comment]
https://jenkins01.arangodb.biz/view/PR/job/arangodb-matrix-pr/7483/

**Comment by ObiWahn** [comment]
https://jenkins01.arangodb.biz/view/PR/job/arangodb-matrix-pr/7487/

File: arangod/RestServer/MetricsFeature.h [hunk 1]
```diff
@@ -34,121 +34,124 @@ namespace arangodb {
...
   template<typename T> Gauge<T>& gauge(std::string const& name) {
-    std::lock_guard<std::mutex> guard(_lock);
-    auto it = _registry.find(name);
+
+    registry_type::const_iterator it;
+    {
+      std::lock_guard<std::mutex> guard(_lock);
+      it = _registry.find(name);
+    }
     if (it == _registry.end()) {
```
Comment (ObiWahn) [inline arangod/RestServer/MetricsFeature.h]: Why is `it` safe to use here?

File: arangod/RestServer/MetricsFeature.h [hunk 2]
```diff
@@ -34,121 +34,124 @@ namespace arangodb {
...
 
   template<typename T> Histogram<T>& histogram (std::string const& name) {
-    std::lock_guard<std::mutex> guard(_lock);
-    auto const it = _registry.find(name);
+
+    registry_type::const_iterator it;
+    {
+      std::lock_guard<std::mutex> guard(_lock);
+      it = _registry.find(name);
+    }
```
Comment (ObiWahn) [inline arangod/RestServer/MetricsFeature.h]: why is it safe to use here?

File: arangod/Statistics/Descriptions.cpp [hunk 3]
```diff
@@ -17,7 +17,7 @@
 ///
 /// Copyright holder is ArangoDB GmbH, Cologne, Germany
 ///
-/// @author Simon Grätzer
+/// @author Simon Graetzer
```
Comment (ObiWahn) [inline arangod/Statistics/Descriptions.cpp]: @graetzer should really correct his name on linkedin :)

File: tests/CMakeLists.txt [hunk 4]
```diff
@@ -246,7 +250,13 @@ add_executable(
   arangodbtests
   ${ARANGODB_TESTS_SOURCES}
   main.cpp
-)
+  )
+
+add_executable(
```
Comment (ObiWahn) [inline tests/CMakeLists.txt]: This should be `EXCLUDE_FROM_ALL` as it is covered with the regular tests above for a default build.

https://cmake.org/cmake/help/v3.16/command/add_executable.html

File: tests/CMakeLists.txt [hunk 5]
```diff
@@ -275,6 +291,10 @@ target_include_directories(arangodbtests PRIVATE
   ${INCLUDE_DIRECTORIES}
 )
 
+target_include_directories(arangodbtests_metrics PRIVATE
```
Comment (ObiWahn) [inline tests/CMakeLists.txt]: How is this library target used by others if it has no PUBLIC includes?

File: arangod/RestServer/MetricsFeature.h [hunk 6]
```diff
@@ -34,121 +34,124 @@ namespace arangodb {
...
-
-    return *h;
-  };
+    return *metric;
+    
+  }
 
   Counter& counter(std::string const& name, uint64_t const& val, std::string const& help);
-  Counter& counter(std::string const& name);
+  //Counter& counter(std::string const& name);
```
Comment (ObiWahn) [inline arangod/RestServer/MetricsFeature.h]: Can this be deleted?

**Comment by ObiWahn** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/7494/

**Comment by ObiWahn** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/7499/

**Comment by ObiWahn** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/7500/

File: arangod/RestServer/MetricsFeature.cpp [hunk 7]
```diff
@@ -111,38 +102,49 @@ void MetricsFeature::toPrometheus(std::string& result) const {
...
-  return *c;
-};
+  if (!success) {
+    THROW_ARANGO_EXCEPTION_MESSAGE(
+      TRI_ERROR_INTERNAL, std::string("counter ") + name + " alredy exists");
+  }
+  return *metric;
+}
 
+/*
```
[the remark this answers is not in the document]
Comment (kvahed) [reply arangod/RestServer/MetricsFeature.cpp]: done

**Comment by kvahed** [comment]
https://jenkins01.arangodb.biz/view/PR/job/arangodb-matrix-pr/7501/

**Comment by jsteemann** [comment]
tests blue

**Comment by KVS85** [comment]
http://jenkins.arangodb.biz:8080/job/arangodb-ANY-examples/471/

**Pull request merged**